internal/olm/operator/registry/operator_installer.go

```diff
@@ -122,48 +124,60 @@ func (o OperatorInstaller) waitForCatalogSource(ctx context.Context, cs *v1alpha
 	return nil
 }
 
-// createOperatorGroup creates an OperatorGroup using package name if an OperatorGroup does not exist.
-// If one exists in the desired namespace and it's target namespaces do not match the desired set,
-// createOperatorGroup will return an error.
-func (o OperatorInstaller) createOperatorGroup(ctx context.Context) error {
-	targetNamespaces := make([]string, len(o.InstallMode.TargetNamespaces), cap(o.InstallMode.TargetNamespaces))
-	copy(targetNamespaces, o.InstallMode.TargetNamespaces)
+func (o OperatorInstaller) ensureOperatorGroup(ctx context.Context) error {
 	// Check OperatorGroup existence, since we cannot create a second OperatorGroup in namespace.
 	og, ogFound, err := o.getOperatorGroup(ctx)
 	if err != nil {
 		return err
 	}
-	// TODO: we may need to poll for status updates, since status.namespaces may not be updated immediately.
-	if ogFound {
-		// targetNamespaces will always be initialized, but the operator group's namespaces may not be
-		// (required for comparison).
-		if og.Status.Namespaces == nil {
-			og.Status.Namespaces = []string{}
+
+	supported := o.SupportedInstallModes
+
+	// --install-mode was given
+	if !o.InstallMode.IsEmpty() {
```

Comment:
Why are we still checking install mode compatibility in this method? This should all be validated by the time `ensureOperatorGroup` is called. Is there a reason why you're checking them here and `CheckCompatibility`?

Reply:
I think I agree @estroz, but sanity check me here @jmrodri. There are three main things we need to check, and I think we can do ALL of it before creating any resources in the cluster:
1. The actual format and content of the `--install-mode` flag. We can check this during flag parsing with `InstallMode.Set()`.
2. CSV / `--install-mode` compatibility:
   - if the `--install-mode` flag is set, make sure it is supported by the CSV.
   - if the `--install-mode` flag is not set, make sure the CSV supports one of `AllNamespaces` or `OwnNamespace` (`SingleNamespace` requires an explicit target namespace from the `--install-mode` flag).
3. There can only be 0 or 1 operator groups in the namespace. If the namespace has an existing operator group, is it compatible with the intersection of the CSV supported install modes and the `--install-mode` flag (if set).

Reply:
@estroz @joelanford you're literally saying do not make these checks in `ensureOperatorGroup`? I realize they are being done in `CheckCompatibility` which is the earliest entrypoint where the users input can be checked. But checking it there does not guarantee that some other code can't call `ensureOperatorGroup` without having gone through `CheckCompatibility`.  Or are we saying that OperatorInstaller does *NOT* care about the rules it only cares about passing the data that was verified BEFORE it was called?

Reply:
I guess it feels weird not doing any checks in `ensureOperatorGroup` because `OperatorInstaller` does not know that all its attributes were set correctly, so it seems like it should ensure they're okay before doing something with them.